Source/WebCore/page/Navigation.cpp

```diff
@@ -127,16 +127,52 @@ static Navigation::Result createErrorResult(Ref<DeferredPromise> committed, Ref<
     return createErrorResult(committed, finished, Exception { exceptionCode, errorMessage });
 }
 
+ExceptionOr<std::optional<Ref<SerializedScriptValue>>> Navigation::serializeState(JSC::JSValue& state)
+{
+    if (state.isUndefined())
+        return { std::nullopt };
+
+    Vector<RefPtr<MessagePort>> dummyPorts;
+    auto serializeResult = SerializedScriptValue::create(*protectedScriptExecutionContext()->globalObject(), state, { }, dummyPorts, SerializationForStorage::Yes);
+    if (serializeResult.hasException())
+        return serializeResult.releaseException();
+
+    return { serializeResult.releaseReturnValue() };
+}
+
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#maybe-set-the-upcoming-non-traverse-api-method-tracker
+Ref<NavigationAPIMethodTracker> Navigation::maybeSetUpcomingNonTraversalTracker(Ref<DeferredPromise>& committed, Ref<DeferredPromise>& finished, JSC::JSValue&& info, std::optional<Ref<SerializedScriptValue>>&& serializedState)
+{
+    auto globalObject = committed->globalObject();
+    Ref committedPromise = DOMPromise::create(*globalObject, *JSC::jsCast<JSC::JSPromise*>(committed->promise()));
+    Ref finishedPromise = DOMPromise::create(*globalObject, *JSC::jsCast<JSC::JSPromise*>(finished->promise()));
```

Comment:
These should probably be properly moved into NavigationAPIMethodTracker::create or NavigationAPIMethodTracker should create them itself (based on the two DeferredPromise params).